### PR TITLE
PageStorage APIs support throw_on_not_exist.

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: ref #3594 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/PageDefines.h [hunk 1]
```diff
@@ -54,6 +54,7 @@ static constexpr NamespaceId TEST_NAMESPACE_ID = 1000;
 using PageId = UInt64;
 using PageIds = std::vector<PageId>;
 using PageIdSet = std::unordered_set<PageId>;
+static constexpr PageId INVALID_PAGE_ID = std::numeric_limits<PageId>::max();
```
Comment (jiaqizho) [inline dbms/src/Storages/Page/PageDefines.h:57]: change to 0

File: dbms/src/Storages/Page/V2/PageFile.cpp [hunk 2]
```diff
@@ -1070,6 +1069,72 @@ PageMap PageFile::Reader::read(PageFile::Reader::FieldReadInfos & to_read, const
...
+        PageUtil::readFile(data_file, to_read.entry.offset + beg_offset, write_offset, size_to_read, read_limiter);
+        fields_offset_in_page.emplace(field_index, read_size_this_entry);
+
+        if constexpr (PAGE_CHECKSUM_ON_READ)
+        {
+            auto expect_checksum = to_read.entry.field_offsets[field_index].second;
+            auto field_checksum = CityHash_v1_0_2::CityHash64(write_offset, size_to_read);
+            if (unlikely(to_read.entry.size != 0 && field_checksum != expect_checksum))
+            {
+                throw Exception(fmt::format("Page [{}] field [{}] checksum not match, broken file: {}, expected: 0x{:X}, but: 0x{:X}",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V2/PageFile.cpp:1107]: Adding the entry.offset and entry.size to exception message is better

File: dbms/src/Storages/Page/PageStorage.h [hunk 3]
```diff
@@ -179,47 +179,60 @@ class PageStorage : private boost::noncopyable
...
-        return readImpl(ns_id, page_ids, read_limiter, snapshot);
+        return readImpl(ns_id, page_ids, read_limiter, snapshot, throw_on_not_exist);
     }
 
-    void read(NamespaceId ns_id, const std::vector<PageId> & page_ids, const PageHandler & handler, const ReadLimiterPtr & read_limiter = nullptr, SnapshotPtr snapshot = {})
+    /**
+     * If throw_on_not_exist is false, Also we do have some of page_id not found.
+     * Then the return value will record the all of page_id which not found.
+     */
+    PageIds read(NamespaceId ns_id, const PageIds & page_ids, const PageHandler & handler, const ReadLimiterPtr & read_limiter = nullptr, SnapshotPtr snapshot = {}, bool throw_on_not_exist = true)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/PageStorage.h:210]: why do we need this return value for `read(..., page_ids, ...)` but don't need it for other `read`?

File: dbms/src/Storages/Page/V3/PageStorageImpl.cpp [hunk 4]
```diff
@@ -126,18 +131,25 @@ DB::PageEntry PageStorageImpl::getEntryImpl(NamespaceId ns_id, PageId page_id, S
...
+DB::Page PageStorageImpl::readImpl(NamespaceId ns_id, PageId page_id, const ReadLimiterPtr & read_limiter, SnapshotPtr snapshot, bool throw_on_not_exist)
 {
     if (!snapshot)
     {
         snapshot = this->getSnapshot("");
     }
 
-    auto page_entry = page_directory->get(buildV3Id(ns_id, page_id), snapshot);
+    auto page_entry = throw_on_not_exist ? page_directory->get(buildV3Id(ns_id, page_id), snapshot) : page_directory->getOrNull(buildV3Id(ns_id, page_id), snapshot);
+    if (page_entry.second.file_id == INVALID_BLOBFILE_ID)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageStorageImpl.cpp:142]: Add a `isValid` method for `PageEntryV3`?

File: dbms/src/Storages/Page/V3/tests/gtest_page_storage.cpp [hunk 5]
```diff
@@ -122,6 +122,70 @@ try
...
+        batch.putPage(4, 0, std::make_shared<ReadBufferFromMemory>(c_buff, buf_sz), buf_sz, {20, 20, 30, 30});
+        page_storage->write(std::move(batch));
+    }
+
+    {
+        PageIds page_ids = {1, 2, 5};
+        // readImpl(TEST_NAMESPACE_ID, page_ids, nullptr, nullptr, true);
+        auto page_maps = page_storage->readImpl(TEST_NAMESPACE_ID, page_ids, nullptr, nullptr, false);
+        ASSERT_EQ(page_maps[1].page_id, 1);
+        ASSERT_EQ(page_maps[2].page_id, INVALID_PAGE_ID);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_page_storage.cpp:149]: maybe we should add `isValid()` for the class `Page`

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Review by CalvinNeo — APPROVED** [review]
LGTM

**Pull request merged**